concurrency/executor.go

```diff
@@ -0,0 +1,212 @@
+// one use SimpleTaskProcessorFn instead.
+type TaskProcessorFn[ResultChannelType, TaskType any] func(resultCh chan ResultChannelType, errorCh chan error, executorNum int, payload TaskType)
+
+// SimpleTaskProcessorFn is a function that processes a task that doesn't require a payload. It should send the result to the resultCh and any error to the errorCh. It should
+// never send to both channels. The executorNum is the index of the executor that is processing the task.
+type SimpleTaskProcessorFn[ResultChannelType any] func(resultCh chan ResultChannelType, errorCh chan error, executorNum int)
+
+// ChannelWithResult is an interface that should be implemented by the result channel
+type ChannelWithResult[ResultType any] interface {
+	GetResult() ResultType
+}
+
+// GetResults returns all errors that occurred during processing
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) GetErrors() []error {
+	return e.errors
+}
+
+// ExecuteSimple executes a task that doesn't require a payload. It is executed repeatTimes times with given concurrency. The simpleProcessorFn is the function that processes the task.
+// Executor will attempt to distribute the tasks evenly among the executors.
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) ExecuteSimple(concurrency int, repeatTimes int, simpleProcessorFn SimpleTaskProcessorFn[ResultChannelType]) ([]ResultType, error) {
+	dummy := make([]TaskType, repeatTimes)
+	for i := 0; i < repeatTimes; i++ {
+		dummy[i] = *new(TaskType)
+	}
+
+	return e.Execute(concurrency, dummy, adaptSimpleToTaskProcessorFn[ResultChannelType, TaskType](simpleProcessorFn))
+}
+
+// Execute executes a task that requires a payload. It is executed with given concurrency. The processorFn is the function that processes the task.
+// Executor will attempt to distribute the tasks evenly among the executors.
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) Execute(concurrency int, payload []TaskType, processorFn TaskProcessorFn[ResultChannelType, TaskType]) ([]ResultType, error) {
+	if len(payload) == 0 {
+		return []ResultType{}, nil
+	}
+
+	if concurrency <= 0 {
+		e.logger.Warn().Msg("Concurrency is less than 1, setting it to 1")
+		concurrency = 1
+	}
+
+	var wgProcesses sync.WaitGroup
+	wgProcesses.Add(len(payload))
+
+	canSafelyContinueCh := make(chan struct{}) // waits until listening goroutine finishes, so we can safely return from the function
+	doneProcessingCh := make(chan struct{})    // signals that both result and error channels are closed
+	errorCh := make(chan error, len(payload))
+	resultCh := make(chan ResultChannelType, len(payload))
+
+	// mutex to protect shared state
+	mutex := sync.Mutex{}
+
+	// atomic counter to keep track of processed tasks
+	var atomicCounter atomic.Int32
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
I didn't want to do it as none of the methods in core that uses it has context... wdyt about adding it to the struct and using if present, otherwise using a new one?